Improve BytesRef creation from String

schlosna opened the pull request:
### Description

<!--
If this is your first contribution to Lucene, please make sure you have reviewed the contribution guide.
https://github.com/apache/lucene/blob/main/CONTRIBUTING.md
-->

When creating a `BytesRef` from a String, we can leverage `String::getBytes(StandardCharset.UTF_8)` to avoid 3x memory allocation overhead when the `String` is already `UTF-8` and when the [JDK provides compact `String` representation as `byte[]`](https://openjdk.org/jeps/254) this becomes a single array copy rather than more complex
`org.apache.lucene.util.UnicodeUtil#UTF16toUTF8` logic. Benchmark results for the added `BytesRefBenchmark` on MacBookPro M2 Max show specialized `BytesRef(String)` constructor provides 2x to 6x throughput improvement compared to `BytesRef(CharSequence)` constructor, and 3x less allocations for ASCII Strings.

```
# Apple MacBookPro M2 Max
# JMH version: 1.37
# VM version: JDK 24.0.1, OpenJDK 64-Bit Server VM, 24.0.1+9-FR
# VM invoker: /Library/Java/JavaVirtualMachines/amazon-corretto-24.jdk/Contents/Home/bin/java
# VM options: -XX:+UseParallelGC -XX:TieredStopAtLevel=1 -XX:ActiveProcessorCount=1 -Dfile.encoding=UTF-8 -Duser.country=US -Duser.language=en -Duser.variant
# Blackhole mode: compiler (auto-detected, use -Djmh.blackhole.autoDetect=false to disable)
# Warmup: 3 iterations, 3 s each
# Measurement: 5 iterations, 3 s each

                                                                         (before)              (after)
                                                                       bytesRefCharSequence   bytesRefString
Benchmark                             (length)      (type)   Mode  Cnt      Score     Error      Score     Error   Units
BytesRefBenchmark                         1000       ASCII  thrpt    5      1.213 ±   0.022      6.090 ±   0.145  ops/us
BytesRefBenchmark:gc.alloc.rate.norm      1000       ASCII  thrpt    5   3040.002 ±   0.001   1040.000 ±   0.001    B/op
BytesRefBenchmark                         1000  ISO_8859_1  thrpt    5      1.123 ±   0.022      3.334 ±   0.116  ops/us
BytesRefBenchmark:gc.alloc.rate.norm      1000  ISO_8859_1  thrpt    5   3040.002 ±   0.001   3552.001 ±   0.001    B/op
BytesRefBenchmark                         1000   UTF_8_BMP  thrpt    5      0.660 ±   0.017      1.548 ±   0.092  ops/us
BytesRefBenchmark:gc.alloc.rate.norm      1000   UTF_8_BMP  thrpt    5   3040.004 ±   0.001   5968.002 ±   0.001    B/op
BytesRefBenchmark                         1000      UTF_16  thrpt    5      0.435 ±   0.009      0.811 ±   0.068  ops/us
BytesRefBenchmark:gc.alloc.rate.norm      1000      UTF_16  thrpt    5   5888.006 ±   0.001   9848.003 ±   0.001    B/op
```

schlosna committed:
Improve BytesRef creation from String

When creating a BytesRef from a String, we can leverage
String::getBytes(StandardCharset.UTF_8) to avoid potential 3x memory
allocation overhead when the String is already UTF-8 and when JDK
provides compressed String representation of byte[] this becomes a
single array copy rather than more complex
org.apache.lucene.util.UnicodeUtil#UTF16toUTF8 logic.

msfroh commented:
That's a nice improvement! 

Presumably this kicks in when we have [compact strings](https://openjdk.org/jeps/254)? Do you have a JMH benchmark with non-ISO-8859-1/Latin-1 characters?

schlosna commented:
> Presumably this kicks in when we have [compact strings](https://openjdk.org/jeps/254)? Do you have a JMH benchmark with non-ISO-8859-1/Latin-1 characters?

Yes, this leverages compact strings for Latin-1 for the best case scenario here. I'll throw together a more rigorous benchmark including non-Latin1.

schlosna committed:
Add BytesRefBenchmark

Benchmark shows String specialized BytesRef constructor provides 2x to 6x the
throughput compared to CharSequence constructor, and 3x less allocations for
ASCII Strings.

```

                                                                         bytesRefCharSequence        bytesRefString
Benchmark                             (length)      (type)   Mode  Cnt      Score     Error   Units     Score     Error   Units
BytesRefBenchmark                         1000       ASCII  thrpt    5      1.213 ±   0.022  ops/us     6.090 ±   0.145  ops/us
BytesRefBenchmark:gc.alloc.rate.norm      1000       ASCII  thrpt    5   3040.002 ±   0.001    B/op  1040.000 ±   0.001    B/op
BytesRefBenchmark                         1000  ISO_8859_1  thrpt    5      1.123 ±   0.022  ops/us     3.334 ±   0.116  ops/us
BytesRefBenchmark:gc.alloc.rate.norm      1000  ISO_8859_1  thrpt    5   3040.002 ±   0.001    B/op  3552.001 ±   0.001    B/op
BytesRefBenchmark                         1000   UTF_8_BMP  thrpt    5      0.660 ±   0.017  ops/us     1.548 ±   0.092  ops/us
BytesRefBenchmark:gc.alloc.rate.norm      1000   UTF_8_BMP  thrpt    5   3040.004 ±   0.001    B/op  5968.002 ±   0.001    B/op
BytesRefBenchmark                         1000      UTF_16  thrpt    5      0.435 ±   0.009  ops/us     0.811 ±   0.068  ops/us
BytesRefBenchmark:gc.alloc.rate.norm      1000      UTF_16  thrpt    5   5888.006 ±   0.001    B/op  9848.003 ±   0.001    B/op
```

schlosna committed:
changes

schlosna commented:
Added `BytesRefBenchmark` demonstrating existing `new BytesRef(CharSequence)` vs. `new BytesRef(String)` demonstrating 2x to 6x throughput improvement on AMD EPYC 7R13 Processor:

```
# AMD EPYC 7R13 Processor
# JMH version: 1.37
# VM version: JDK 24.0.1, OpenJDK 64-Bit Server VM, 24.0.1+9-FR
# VM invoker: /home/dev/.gradle/jdks/amazon-corretto-24.0.1.9.1-linux-x64/bin/java
# VM options: -XX:+UseParallelGC -XX:TieredStopAtLevel=1 -XX:ActiveProcessorCount=1 -Dfile.encoding=UTF-8 -Duser.country=US -Duser.language=en -Duser.variant
# Blackhole mode: compiler (auto-detected, use -Djmh.blackhole.autoDetect=false to disable)
# Warmup: 3 iterations, 3 s each
# Measurement: 5 iterations, 3 s each

                                                      (before)             (after)
                                                    bytesRefCharSequence   bytesRefString
Benchmark          (length)      (type)   Mode  Cnt      Score     Error     Score     Error   Units
BytesRefBenchmark        10       ASCII  thrpt    5     75.650 ±   4.078   180.382 ±   7.672  ops/us
BytesRefBenchmark        10  ISO_8859_1  thrpt    5     66.688 ±   3.957   114.655 ±   7.510  ops/us
BytesRefBenchmark        10   UTF_8_BMP  thrpt    5     46.678 ±   2.387    77.036 ±   2.484  ops/us
BytesRefBenchmark        10      UTF_16  thrpt    5     26.289 ±   1.376    49.021 ±   1.740  ops/us
BytesRefBenchmark       100       ASCII  thrpt    5      9.106 ±   0.487    47.246 ±   1.218  ops/us
BytesRefBenchmark       100  ISO_8859_1  thrpt    5      8.447 ±   0.628    23.809 ±   0.488  ops/us
BytesRefBenchmark       100   UTF_8_BMP  thrpt    5      5.129 ±   0.141    10.431 ±   0.336  ops/us
BytesRefBenchmark       100      UTF_16  thrpt    5      2.963 ±   0.127     5.904 ±   0.271  ops/us
BytesRefBenchmark      1000       ASCII  thrpt    5      0.917 ±   0.038     5.891 ±   0.428  ops/us
BytesRefBenchmark      1000  ISO_8859_1  thrpt    5      0.830 ±   0.024     1.981 ±   0.077  ops/us
BytesRefBenchmark      1000   UTF_8_BMP  thrpt    5      0.468 ±   0.014     0.950 ±   0.057  ops/us
BytesRefBenchmark      1000      UTF_16  thrpt    5      0.298 ±   0.011     0.537 ±   0.021  ops/us
```

vigyasharma commented on lucene/CHANGES.txt:
```diff
@@ -41,6 +41,7 @@ Optimizations
 * GITHUB#14011: Reduce allocation rate in HNSW concurrent merge. (Viliam Durina)
 * GITHUB#14022: Optimize DFS marking of connected components in HNSW by reducing stack depth, improving performance and reducing allocations. (Viswanath Kuchibhotla)
+* GITHUB#14678: Optimize BytesRef creation from Strings, improving throughput and reducing allocations. (David Schlosnagle)
 
```
This could go in 10.3, any reason to keep this for 11.0?

vigyasharma reviewed (commented):
Nice, impressive ~6x improvement with ASCII strings! 

Is the ~1.8x improvement with `UTF_16` strings also expected? I thought this worked on compact strings that need 1 byte per char?

rmuir commented:
There's intentionally not a `BytesRef(String)` constructor, as there have been issues before.

schlosna commented:
> There's intentionally not a `BytesRef(String)` constructor, as there have been issues before.

@rmuir Do you happen to recall what issues a `BytesRef(String)` constructor create and any history there? 

I imagine you're referring to the case where the provided text is not "well-formed unicode text, with no unpaired surrogates" as defined in the `BytesRef(CharSequence)` constructor, and where `org.apache.lucene.util.UnicodeUtil#validUTF16String(java.lang.CharSequence)` returns false. `UnicodeUtil.UTF16toUTF8` used by `BytesRef(CharSequence)` replaces unpaired surrogates with `�` bytes `0xEF 0xBF 0xBD` while the `BytesRef(String)` constructor as proposed does no replacement or validation.

https://github.com/apache/lucene/blob/b0f992369ba967341d9b51f05f41ab25eb177e1d/lucene/core/src/java/org/apache/lucene/util/BytesRef.java#L77-L81

https://github.com/apache/lucene/blob/b0f992369ba967341d9b51f05f41ab25eb177e1d/lucene/core/src/java/org/apache/lucene/util/UnicodeUtil.java#L168-L172


In https://github.com/apache/lucene/issues/12071 these is mention https://github.com/apache/lucene/issues/12071#issuecomment-1379313710 of using the vector APIs to speed up UnicodeUtil conversions. Has any of that been approached in lucene yet? I imagine a SIMD or SWAR approach like https://lemire.me/blog/2018/05/16/validating-utf-8-strings-using-as-little-as-0-7-cycles-per-byte/ & https://lemire.me/blog/2023/02/16/computing-the-utf-8-size-of-a-latin-1-string-quickly-avx-edition/ would be worth investigating.

rmuir commented:
git log tells the story. A quick summary:

* conflicting use cases such as usages as a reference vs stringbuilder/string type uses
* ownership issues and API problems resulting from that.

The use-cases were torn apart and the code cleaned up and bytesrefbuilder made separate. we shouldn't go backwards.

this is basically a reference class, intended to support pointing to byte[]s held elsewhere. like a pointer in c lol.

rmuir commented:
> In #12071 these is mention [#12071 (comment)](https://github.com/apache/lucene/issues/12071#issuecomment-1379313710) of using the vector APIs to speed up UnicodeUtil conversions. Has any of that been approached in lucene yet? I imagine a SIMD or SWAR approach like https://lemire.me/blog/2018/05/16/validating-utf-8-strings-using-as-little-as-0-7-cycles-per-byte/ & https://lemire.me/blog/2023/02/16/computing-the-utf-8-size-of-a-latin-1-string-quickly-avx-edition/ would be worth investigating.

If you want to speed something up, pay no mind to java.lang.String as it isn't involved in the indexing anyway. Look here which is what the indexer calls jazillions of times https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/analysis/tokenattributes/CharTermAttributeImpl.java#L91-L95

it is just char[]->byte[] conversion, with the same reused char[] and byte[], there aren't allocations or strings or any of that.

schlosna commented:
> > In #12071 these is mention [#12071 (comment)](https://github.com/apache/lucene/issues/12071#issuecomment-1379313710) of using the vector APIs to speed up UnicodeUtil conversions. Has any of that been approached in lucene yet? I imagine a SIMD or SWAR approach like https://lemire.me/blog/2018/05/16/validating-utf-8-strings-using-as-little-as-0-7-cycles-per-byte/ & https://lemire.me/blog/2023/02/16/computing-the-utf-8-size-of-a-latin-1-string-quickly-avx-edition/ would be worth investigating.
> 
> If you want to speed something up, pay no mind to java.lang.String as it isn't involved in the indexing anyway. Look here which is what the indexer calls jazillions of times https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/analysis/tokenattributes/CharTermAttributeImpl.java#L91-L95
> 
> it is just char[]->byte[] conversion, with the same reused char[] and byte[], there aren't allocations or strings or any of that.

Thanks, I'm going to close out this PR for now. When I have a chance I will take a look at optimizing the `char[]` to `byte[]` conversions via `jdk.incubator.vector.ShortVector.fromCharArray`.

rmuir commented:
yeah it is tricky, since the 'strings' indexed in search are usually small: words. for a lot of natural languages average word length is already small (e.g. english: ~5), and often in search people are shortening that even more (with stemmers etc).
